oak_crypto/src/noise_nk/noise.rs

```diff
@@ -0,0 +1,252 @@
+    }
+
+    pub fn mix_key(&mut self, ikm: &[u8]) {
+        // See https://www.noiseprotocol.org/noise.html#the-symmetricstate-object
+        let derived_keys = hkdf2(&self.chaining_key, ikm);
+        self.chaining_key.copy_from_slice(&derived_keys.0);
+        self.initialize_key(&derived_keys.1);
+    }
+
+    #[cfg(test)]
+    pub fn mix_key_and_hash(&mut self, ikm: &[u8]) {
+        // See https://www.noiseprotocol.org/noise.html#the-symmetricstate-object
+        let mut output = [0; SYMMETRIC_KEY_SIZE * 3];
+        // unwrap: only fails if the output size is too large, but the output
+        // size is small and fixed here.
+        rustcrypto::hkdf_sha256(ikm, &self.chaining_key, &[], &mut output).unwrap();
+        self.chaining_key.copy_from_slice(&output[..SYMMETRIC_KEY_SIZE]);
+        self.mix_hash(&output[SYMMETRIC_KEY_SIZE..SYMMETRIC_KEY_SIZE * 2]);
+        self.initialize_key(&output[SYMMETRIC_KEY_SIZE * 2..].try_into().unwrap());
+    }
+
+    fn next_nonce(&mut self) -> [u8; NONCE_LEN] {
+        let mut nonce_bytes = [0; NONCE_LEN];
+        nonce_bytes[0] = (self.symmetric_nonce >> 24) as u8;
+        nonce_bytes[1] = (self.symmetric_nonce >> 16) as u8;
+        nonce_bytes[2] = (self.symmetric_nonce >> 8) as u8;
+        nonce_bytes[3] = self.symmetric_nonce as u8;
+        self.symmetric_nonce += 1;
+        nonce_bytes
+    }
+
+    pub fn encrypt_and_hash(&mut self, plaintext: &[u8]) -> Vec<u8> {
+        let mut encrypted_data = Vec::from(plaintext);
+        let nonce = self.next_nonce();
+        rustcrypto::aes_256_gcm_seal_in_place(
+            &self.symmetric_key,
+            &nonce,
+            &self.h,
+            &mut encrypted_data,
+        );
+        self.mix_hash(&encrypted_data);
+        encrypted_data
+    }
+
+    pub fn decrypt_and_hash(&mut self, ciphertext: &[u8]) -> Result<Vec<u8>, Error> {
+        let h = self.h;
+        self.mix_hash(ciphertext);
+
+        let ciphertext = Vec::from(ciphertext);
+        let nonce = self.next_nonce();
+        let plaintext =
+            rustcrypto::aes_256_gcm_open_in_place(&self.symmetric_key, &nonce, &h, ciphertext)
+                .map_err(|_| Error::DecryptFailed)?;
+        Ok(plaintext)
+    }
+
+    pub fn handshake_hash(&self) -> [u8; SYMMETRIC_KEY_SIZE] {
+        self.h
+    }
+
+    // |point| is the contents from either agreement::PublicKey::as_ref() or
+    // agreement::UnparsedPublicKey::bytes().
+    pub fn mix_hash_point(&mut self, point: &[u8]) {
+        self.mix_hash(point);
+    }
+
+    pub fn traffic_keys(&self) -> ([u8; SYMMETRIC_KEY_SIZE], [u8; SYMMETRIC_KEY_SIZE]) {
+        hkdf2(&self.chaining_key, &[])
+    }
+
+    fn initialize_key(&mut self, key: &[u8; SYMMETRIC_KEY_SIZE]) {
+        // See https://www.noiseprotocol.org/noise.html#the-cipherstate-object
+        self.symmetric_key.copy_from_slice(key);
+        self.symmetric_nonce = 0;
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
nit: also could be moved to a separate file

Reply:
I'm still a noob to Rust, so please correct me if my thinking here is wrong...

I think I prefer to leave it here, as it accesses private functions in the module, which is appropriate for unit tests.  For unit tests, I think in the same file makes sense, and I see that elsewhere in the Oak code base.  If I factored this out, would it be called noise_test.rs?  I see no similar test files in the code base.

In contrast, the test in mod.rs is a sub-module level test and having it use only the sub-module level APIs (which is the case here) makes sense.  Factoring it out to tests.rs shows users how to use the sub-module, including what use statements are required.  I see several tests.rs files for crate-level or sub-module-level testing.